### PR TITLE
Fix master slave reader panic in concurrent

### PR DESCRIPTION
**Pull request by xxx7xxxx**
Fix https://github.com/megaease/easegress/issues/407

### CODE REVIEW

**Review by samutamm — APPROVED** [review]
This seems to fix https://github.com/megaease/easegress/issues/407. LGTM

**Review by suchen-sci — APPROVED** [review]
LGTM!

**Comment by localvar** [comment]
I think this PR could eliminate the issue, but seems the fix is not for the root cause. I'd like to take some time to understand more details about #407 before approval.

**Comment by suchen-sci** [comment]
@localvar Yeah, there are two possible reasons that cause this problem. 1. more than one goroutines call `Read` function. 2. one goroutine call `Read` even `Read` return error. But after carefully check the code, I don't find such places. Another possible place to cause this maybe in stdlib.... Please let me known if you have any idea about this....

**Review by megaeasex — CHANGES_REQUESTED** [review]
[TASK:easegress-PR-test FAILED]megaease/easegress Pull Request 408 Deploy Test failed

File: pkg/filter/proxy/masterslavereader.go [hunk 1]
```diff
@@ -66,7 +68,9 @@ func (mr *masterReader) Read(p []byte) (n int, err error) {
 	}
 
 	if err == io.EOF {
-		close(mr.buffChan)
+		mr.closeChanOnce.Do(func() {
+			close(mr.buffChan)
+		})
```
Comment (suchen-sci) [inline pkg/filter/proxy/masterslavereader.go:73]: If non-EOF error happens, do we need to take care of `mr.buffChan`?

**Comment by qdongxu** [comment]
f```
unc (t *teeReader) Read(p []byte) (n int, err error) {
   n, err = t.r.Read(p)
   if n > 0 {
      if n, err := t.w.Write(p[:n]); err != nil {
         return n, err
      }
   }
   return
}

this is probably the cause, the teeReader returns an error only when n > 0.  It ignores the err = err.EOF  when n == 0.

```
> @localvar Yeah, there are two possible reasons that cause this problem. 1. more than one goroutines call `Read` function. 2. one goroutine call `Read` even `Read` return error. But after carefully check the code, I don't find such places. Another possible place to cause this maybe in stdlib.... Please let me known if you have any idea about this....

File: pkg/filter/proxy/masterslavereader.go (hunk 1 above)
Comment (xxx7xxxx) [reply pkg/filter/proxy/masterslavereader.go:73]: No

**Comment by xxx7xxxx** [comment]
Fixed by another PR https://github.com/megaease/easegress/pull/411

**Pull request closed without merging**